src/orion/core/worker/primary_algo.py

```diff
@@ -21,39 +33,67 @@ class SpaceTransformAlgoWrapper:
 
     Parameters
     ----------
-    algo_constructor: Child class of `BaseAlgorithm`
-        Class constructor to build the algorithm object.
+    algorithm: instance of `BaseAlgorithm`
+        Algorithm to be wrapped.
     space : `orion.algo.space.Space`
        The original definition of a problem's parameters space.
     algorithm_config : dict
        Configuration for the algorithm.
 
     """
 
-    def __init__(self, algo_constructor, space, **algorithm_config):
+    def __init__(self, algorithm: AlgoType, space: Space):
         self._space = space
-        requirements = backward.get_algo_requirements(algo_constructor)
-        self.transformed_space = build_required_space(space, **requirements)
-        self.algorithm = algo_constructor(space, **algorithm_config)
-        self.algorithm.space = self.transformed_space
+        self.algorithm: AlgoType = algorithm
+        self.registry = Registry()
+        self.registry_mapping = RegistryMapping(
+            original_registry=self.registry,
+            transformed_registry=self.algorithm.registry,
+        )
+
+    @property
+    def original_space(self) -> Space:
+        """The original space (before transformations).
+        This is exposed to the outside, but not to the wrapped algorithm.
+        """
+        return self._space
+
+    @property
+    def transformed_space(self) -> TransformedSpace:
+        """The transformed space (after transformations).
+        This is only exposed to the wrapped algo, not to classes outside of this.
+        """
+        return self.algorithm.space
 
     def seed_rng(self, seed):
         """Seed the state of the algorithm's random number generator."""
         self.algorithm.seed_rng(seed)
 
     @property
-    def state_dict(self):
+    def state_dict(self) -> dict:
         """Return a state dict that can be used to reset the state of the algorithm."""
-        return self.algorithm.state_dict
+        # TODO: There's currently some duplicates between:
+        # - self.registry_mapping.original_registry and self.registry
+        # - self.registry_mapping.transformed_registry and self.algorithm.registry
+        return copy.deepcopy(
+            {
+                "algorithm": self.algorithm.state_dict,
+                "registry": self.registry.state_dict,
+                "registry_mapping": self.registry_mapping.state_dict,
```

Comment:
I wounder if we should not recreate the registry_mapping based on the `_mapping`. The original registry and the transformed registry are handling their state_dict separately, so this should not be part of the state_dict of the registry_mapping. So set_state here could be simply `self.registry_mapping = RegistryMapping(self.registry, self.algorithm.registry, mapping=state_dict['registry_mapping'])` or something like that. What do you think?

Reply:
Hmm, interesting. I think it doesn't really hurt to call `set_state` a few extra times than necessary.
The alternative would be to make it so `RegistryMapping.set_state` only saves/restores `self._mapping`, and just trust that the registries themselves are always saved/restored at the same time.

Reply:
Not sure I understand. Do you want the RegistryMapping to save-restore the original and transformed registries? or only `_mapping` ?

Reply:
Only `_mapping`. Other option is to not support `set_state` and force the creation of a new `RegistryMapping(state=state_dict['registry_mapping'])`. Where `state_dict['registry_mapping'] = self.registry_mapping.state_dict`.

Reply:
Fixed in 1be4bfa21b3565f345ebcab94140d31ecc7278f5 (not pushed atm)